### ISSUE TITLE
KV Cache Implementation for Efficient Transformer Inference

### ISSUE THREAD

**Issue by GarthDB** [GarthDB/metal-candle#30]
### Problem
Current transformer inference recomputes key/value projections for all previous tokens at each generation step. This is extremely inefficient for autoregressive generation (e.g., text generation).

### Performance Impact
- **Without KV Cache**: O(n²) computation for sequence length n
- **With KV Cache**: O(n) computation - **10-50x speedup** for long sequences
- Critical for production LLM inference

### Proposed Solution
Implement KV cache in `metal-candle/src/models/cache.rs`:

```rust
/// Cache for key/value tensors in transformer attention
pub struct KVCache {
    /// Cached keys per layer: Vec<(batch, num_heads, seq_len, head_dim)>
    keys: Vec<Tensor>,
    /// Cached values per layer: Vec<(batch, num_heads, seq_len, head_dim)>
    values: Vec<Tensor>,
    /// Current sequence position
    position: usize,
}

impl KVCache {
    pub fn new(num_layers: usize, device: &Device) -> Self;
    
    /// Append new key/value for current token
    pub fn append(&mut self, layer_idx: usize, key: Tensor, value: Tensor) -> Result<()>;
    
    /// Get cached keys/values for a layer
    pub fn get(&self, layer_idx: usize) -> Result<(&Tensor, &Tensor)>;
    
    /// Clear cache for new sequence
    pub fn reset(&mut self);
}
```

### Integration with Qwen Model
Update `metal-candle/src/models/qwen.rs`:

```rust
impl Qwen {
    /// Forward pass with optional KV cache
    pub fn forward(
        &self,
        input_ids: &Tensor,
        attention_mask: Option<&Tensor>,
        kv_cache: Option<&mut KVCache>, // Add cache parameter
    ) -> Result<Tensor>
}

impl QwenDecoderLayer {
    /// Forward pass with KV cache
    pub fn forward(
        &self,
        hidden_states: &Tensor,
        attention_mask: Option<&Tensor>,
        kv_cache: Option<&mut KVCache>,
        layer_idx: usize,
    ) -> Result<Tensor>
}
```

### Usage Example
```rust
let model = Qwen::new(&config, vb)?;
let mut kv_cache = KVCache::new(model.num_layers(), &device);

// First token: no cache
let logits = model.forward(&input_ids, None, Some(&mut kv_cache))?;

// Subsequent tokens: use cache (much faster!)
for _ in 0..max_tokens {
    let next_token = sample(&logits);
    let logits = model.forward(&next_token, None, Some(&mut kv_cache))?;
}
```

### Expected Performance
- **Inference Speedup**: 10-50x for sequences > 100 tokens
- **Memory**: Slight increase (cache storage), but amortized over speed gain
- **Generation Quality**: Identical output (mathematical equivalence)

### Implementation Considerations
1. **Static vs Dynamic**: Start with dynamic (Vec-based), optimize later
2. **Metal Optimization**: Leverage Metal's efficient tensor concatenation
3. **Batch Support**: Handle batched inference (batch dimension in cache)
4. **Attention Mask**: Update mask for cached positions

### Use Cases
- **Ferris AI Assistant**: Fast Rust code generation (128-512 tokens)
- **Chatbots**: Multi-turn conversations with context
- **Long-form Generation**: Documents, articles, code files
- **Any Autoregressive Model**: GPT, LLaMA, Qwen, etc.

### Priority
**High** - Essential for production inference performance

### References
- [HuggingFace KV Cache](https://huggingface.co/docs/transformers/main/en/kv_cache)
- [Fast Transformer Decoding](https://arxiv.org/abs/2211.05102)
- [LLaMA.cpp KV Cache](https://github.com/ggerganov/llama.cpp)

### Related Issues
- Complements #29 (sampling strategies)
- Prerequisite for efficient text generation API

---

**Motivation**: Currently generating text token-by-token in Ferris is slow without cache. This is a standard optimization in all production LLM inference systems.
